fem/fe.cpp

```diff
@@ -278,6 +294,95 @@ void ScalarFiniteElement::ScalarLocalInterpolation(
    }
 }
 
+const DofToQuad &ScalarFiniteElement::GetDofToQuad(const IntegrationRule &ir,
+                                                   DofToQuad::Mode mode) const
+{
+   MFEM_VERIFY(mode == DofToQuad::FULL, "invalid mode requested");
+
+   for (int i = 0; i < dof2quad_array.Size(); i++)
+   {
+      const DofToQuad &d2q = *dof2quad_array[i];
+      if (d2q.IntRule == &ir && d2q.mode == mode) { return d2q; }
+   }
+
+   DofToQuad *d2q = new DofToQuad;
+   const int nqpt = ir.GetNPoints();
+   d2q->FE = this;
+   d2q->IntRule = &ir;
+   d2q->mode = mode;
+   d2q->ndof = Dof;
+   d2q->nqpt = nqpt;
+   d2q->B.SetSize(nqpt*Dof);
+   d2q->Bt.SetSize(Dof*nqpt);
+   d2q->G.SetSize(nqpt*Dim*Dof);
+   d2q->Gt.SetSize(Dof*nqpt*Dim);
+#ifdef MFEM_THREAD_SAFE
+   Vector c_shape(Dof);
+   DenseMatrix vshape(Dof, Dim);
+#endif
+   for (int i = 0; i < nqpt; i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      CalcShape(ip, c_shape);
+      for (int j = 0; j < Dof; j++)
+      {
+         d2q->B[i+nqpt*j] = d2q->Bt[j+Dof*i] = c_shape(j);
+      }
+      CalcDShape(ip, vshape);
+      for (int d = 0; d < Dim; d++)
+      {
+         for (int j = 0; j < Dof; j++)
+         {
+            d2q->G[i+nqpt*(d+Dim*j)] = d2q->Gt[j+Dof*(i+nqpt*d)] = vshape(j,d);
+         }
+      }
+   }
+   dof2quad_array.Append(d2q);
+   return *d2q;
+}
+
+// protected method
+const DofToQuad &ScalarFiniteElement::GetTensorDofToQuad(
```

Comment:
I think we don't need the `mode` since the `mode` is already included in the name of this function and of `GetDofToQuad`